README.md

# Rotabox
*version 0.13.0*
#### A kivy widget with revolving collision detection, custom bounds and multitouch interactivity.

![example](examples/example.gif)

Rotabox is a *kivy widget* with customizable 2D bounds that follow its rotation.  
The users can shape their own, specific bounds, to fit an image (or a series of images in an animation), using  a visual editor *(See Rotaboxer below)*.

Essentially, it's a widget that, while rotating, is able to track specific points on its visual and it can also be used just for that.

Rotabox also offers multitouch interactivity (drag, rotation and scaling).
___
### Compatibility
 python 2.7 + kivy 1.10 - python 3.7 + kivy 2.1

> **Note:** *Rotabox* is being developed on Windows and hasn't been tested on other platforms.


## Features & particularities

### Collision detection methods
 Rotabox offers two collision approaches. 
 They can't be both used at the same time on the same widget and, normally, collisions are thought to happen between widgets that use the same detection method. 
 Combinations between the two are possible but more expensive.
 
* Segment intersection detection (Default method):  
    (See [Introduction to Algorithms 3rd Edition](https://mitpress.mit.edu/books/introduction-algorithms) (ch.33 Computational Geometry)  
    and ['Line Segment Intersection' lecture notes by Jeff Erickson] (http://jeffe.cs.illinois.edu/teaching/373/notes/x06-sweepline.pdf))
    * Supports open-shaped bounds.
    * Interacts with Rotaboxes that use either collision method (more expensive if method is different) and regular widgets.
    * In a positive check against a Rotabox of the same method, instead of *True*, both the intersected sides' indices and their respecrive polygons' indices are returned, in the form of [this_pol_index, this_side_index, that_pol_index, that_side_index].

* Point membership in polygon detection:  
    (See [Even-odd rule](https://en.wikipedia.org/wiki/Even%E2%80%93odd_rule "")) 
    * It can be faster when dealing with complex shapes, as it can benefit from breaking these shapes into more simple ones when making the bounds in the editor.
    * Requires mutual collision checks (Both parties should check for an accurate reading).
    * Interacts with Rotaboxes that use the same collision method (and regular widgets but behaving, itself, like a regular widget while doing so).
    * In a positive check against a Rotabox of the same method, instead of *True*, the checker's collided polygon's index is returned, in a tuple (i) to always evaluate to True.
	
### Open collision bounds (Segment method only)
 If a polygon is open, the segment between the last and first points of the polygon is not considered in the collision checks.
 Since the segment collision method is only concerned with the polygon's sides, a widget can 'enter' an open polygon, passing through the opening, and then hit the back wall from inside, for example.
 Note that *collide_point* doesn't work for an open polygon (i.e. an open polygon cannot be touched).

### Visual point tracking
 Since a rotating widget doesn't really rotate, its points lose their reference to its visual (Positional properties like *top* or *center* don't rotate).
 Rotabox can track any of its own points while rotating, provided that they are predefined (Hence, the custom bounds' ability).
 They then can be accessed using their indices.
 This can be useful, for example, in changing the point of rotation while the widget is rotating.

### Touch interactivity 
 Due to the differences between the Scatter and Rotabox concepts, a way to combine the two couldn't be found. 
 So, Rotabox uses the Scatter widget's code, modified to act on the actual size and position of the widget and child (essential for accurate collision detection). 
 It supports single and multitouch drag, rotation and scaling (the latter two use the *origin* property in the singletouch option).
	
### Cython option
 Rotabox will try to use a compiled cython/c module (*cybounds.so* or *cybounds.pyd*), for an about X5 speedup.
 The module needs to be compiled for specific platforms, using the provided *cybounds.c* and *setup.py* files, and be put where *rotabox.py* is (more info in *cythons* folder).

### Restrictions
* In order to be able to maintain any arbitrary aspect ratio (e.g. its image's ratio), Rotabox can't use the *size_hint* property.  

* Rotabox can only have one child.  
 It's meant to be an *Image* but not necessarily.  
 Grandchildren, however, can collide independently, only if the widget is not rotated ( *angle* must be *0* ).


_______
# API

## Basic Usage
To use Rotabox, just include *rotabox.py* in your project files.

```python
    from rotabox import Rotabox
    ...
	rb = Rotabox()
	rb.add_widget(Image(source='img.png'))
	self.add_widget(rb)
```
The instance's default bounds will be a rectangle, the size of the image, that rotates with it.

## Basics

**angle** *NumericProperty* (0):  
 The angle of rotation in degrees.

**origin** *AliasProperty*:  
 Sets the point of rotation. Default position is the widget's center.

**image** *ObjectProperty*:  
 Rotabox's only child will most likely be an *Image*.  
 If not so, Rotabox will attempt to find the topmost Image in its tree and assign it to this property by default.  
 Otherwise, the user can specify an *image* somewhere in the widget's tree, that the custom bounds will use as a reference.  
 An .atlas spritesheet can also be used as an animation source and different bounds can be defined for each frame.

**aspect_ratio** *NumericProperty* (0.)  
 If not provided, *image*'s ratio is going to be used.

## Customizing the Collidable Area

> **Rotaboxer** Visual editor.
>  An easy way to define the *custom_bounds* of Rotabox.  
>  To use it, run *rotaboxer.py* directly. It can be found in the *visual_editor* folder.  
>  Open a *.png* image or an *.atlas* file, let the editor draw bounds for it while you control the accuracy/complexity, inspect for a final touch and export the resulting code to clipboard, to paste in a Rotabox widget.
> 
> ![editor](examples/editor.gif)

**custom_bounds** *ObjectProperty* (`[[(0, 0), (1, 0), (1, 1), (0, 1)]]`)  
 This is where the custom bounds are being defined.  
 It's also the output of the Rotaboxer tool ( *above* ).
 It can be a *list* of one or more polygons' data as seen in its default value, above. 
 
 Each polygon's data is a *list* of point tuples `(x, y)`.  
 Points' values should be expressed as percentages of the widget's *width* and *height*, where `(0, 0)` is widget's `(x, y)`,  
 `(1, 1)` is widget's `(right, top)` and `(.5, .5)` is widget's *center*.
 
 Here's another example with more polygons:

```python
rb.custom_bounds = [[(0.013, 0.985), (0.022, 0.349),
                     (0.213, 0.028), (0.217, 0.681)],
                    [(0.267, 0.346), (0.483, -0.005),
                      (0.691, 0.316), (0.261, 0.975)],
                    [(0.539, 0.674), (0.73, 0.37),
                      (0.983, 0.758)]]
```

*custom_bounds* can also be a *dictionary*, in case of animated bounds (different bounds for different frames of an animation sequence in an *.atlas* file), where the *keys* correspond to the frame names in the *.atlas* file and each *item* is a *list* of one or more polygons' data like the above.
  
Here's an example of such a *dictionary*:

```python
rb.custom_bounds = {'00': [[(0.201, 0.803), (0.092, 0.491),
                            (0.219, 0.184), (0.526, 0.064)],
                           [(0.419, 0.095), (0.595, 0.088),
                             (0.644, 0.493)]],
                    '01': [[(0.357, 0.902), (0.17, 0.65),
                            (0.184, 0.337), (0.343, 0.095),
                            (0.644, 0.098)]],
                    '02': [[(...
                             ...
                             ... etc ]]}
```

**segment_mode** *BooleanProperty* (True):
 Toggle between the two collision detection methods *(See Features above)*.
    
**open_bounds** *ListProperty*:
 If a polygon's index is in this list, the segment between the last and first points of the polygon is not considered in the collision checks (segment_mode only).

**pre_check** *BooleanProperty* (False):
 A collision optimization switch for larger widgets in Cython.
 It's always True in Python but in Cython, for small widgets (under 45 points), the slight tax of extra calculations outweighs any benefit in collision.

## Touch interface
Most of it is familiar from the Scatter widget.

**touched_to_front** *BooleanProperty* (False)
 If touched, the widget will be pushed to the top of the parent's widget tree.

**collide_after_children** *BooleanProperty* (True)
 If True, limiting the touch inside the bounds will be done after dispaching the touch to the child and grandchildren, so even outside the bounds they can still be touched.
*IMPORTANT NOTE: Grandchildren, inside or outside the bounds, can collide independently ONLY if widget is NOT ROTATED ( *angle* must be *0* ).*

### Single touch definitions:
**single_drag_touch** *BoundedNumericProperty* (1, min=1)
 How many touches will be treated as one single drag touch.

**single_trans_touch** *BoundedNumericProperty* (1, min=1)
 How many touches will be treated as one single transformation touch.

### Single touch operations:
**allow_drag_x** *BooleanProperty* (False)

**allow_drag_y** *BooleanProperty* (False)

**allow_drag** *AliasProperty*

**single_touch_rotation** *BooleanProperty* (False)
 Rotate around *origin*.

**single_touch_scaling** *BooleanProperty* (False)
 Scale around *origin*.

### Multitouch rotation/scaling:
**multi_touch_rotation** *BooleanProperty* (False)

**multi_touch_scaling** *BooleanProperty* (False)


## Utility interface

**pivot** *ReferenceListProperty*
 The point of rotation and scaling.
 While *origin* property sets *pivot*'s position, relatively to widget's *size* and *pos*, *pivot* itself can be used to position the widget, much like *pos* or *center*.

**ready** *BooleanProperty* (False)
 Useful to read in cases where the widget is stationary. 
 Signifies the completion of the widget's initial preparations.
 Also, its state changes to True after every size change or reset.
 
**prepared** *BooleanProperty* (False)
 Its state change signifies a reset.
 The reset completion signal, however, is the consequent *ready* state change to True.
 
**get_point(pol_index, point_index)** *Method*
 Returns the current position of a certain point.
 The argument indices are based on *custom_bounds*' structure.
 
**read_bounds(filename)** *Method*
 Define *custom_bounds* using a Rotaboxer's project file.
 (e.g. self.custom_bounds = self.read_ bounds("images/car.bounds")
 To work, *size* should be already defined.
 
**draw_bounds** *NumericProperty* (0)
 This option can be useful during testing, as it makes the widget's bounds visible.
 (1 for bounds, 2 for bounds & bounding boxes)
 
**scale** *AliasProperty*
 Current widget's scale, based on widget's original size (User's initial *size* input or *image*'s *texture_size* ).

**scale_min** *NumericProperty* (0.01)

**scale_max** *NumericProperty* (1e20)
 Optional scale restrictions.



_



change_log.md
### Rotabox 0.13.7 changes

##### Changed
* Minor changes


### Rotabox 0.13.6 changes

##### Changed
* `Image.keep_data` fix correction in editor


### Rotabox 0.13.4 changes

##### Changed
* Minor changes in editor and documentation
* Added a prebuilt module for Python 3.7


### Rotabox 0.13.2 changes

##### Changed
* kivy 2.0 compatibility
* Fixed problem with `Image.keep_data`.


### Rotabox 0.13.1 changes

##### Changed
* Minor corrections and changes.


### Rotabox 0.13.0 changes

##### Added

* #### in Rotaboxer (bounds editor):
  * Automated bounding process, using the image's outline (it's detected with alpha/color-tracing).  
    Controls:  
    * complexity/accuracy adjustment  
    * choice of the background color needed to trace .atlas files  
    * option for images with more than one independent shapes   
    * option for processing an entire .atlas animation,  in one go

  
##### Changed
* Minor corrections and changes.


### Rotabox 0.12.1 changes

##### Changed
* Rotabox is now Python 3 compatible.
* Rotaboxer is now Python 3 compatible.


### Rotabox 0.12.0 changes

##### Added
* Cython option:  
Rotabox tries by default to use a compiled cython module (cybounds.so or cybounds.pyd) for an about X5 speedup.
User needs to compile it for specific operating systems using the provided cybounds.c file.

* [pre_check] BooleanProperty(False):
A collision optimization switch for larger widgets in Cython.
It's always considered True in Python but in Cython, for small widgets (under 45 points), the slight tax of the extra calculations outweighs any benefits in collision.

* [read_bounds] A method to define [custom_bounds] by reading a Rotaboxer project file (.bounds file), e.g.: 
```python
self.custom_bounds = self.read_ bounds("images/car.bounds")
```
* #### in Rotaboxer (bounds editor):
  * Multiple points can be moved simultaneously with keyboard arrow keys.
  * Point multiselection (Ctrl+click) was already there since version 0.9 for transfering points between polygons.

##### Removed
* [hidden_bounds] were deemed an unnessesary complication in the recent refactoring, adding too much code for a mere application of a more general feature:
  The ability to treat specific collisions specifically, using multiple polygons and Rotabox' collision feedback.

##### Changed
* Major refactoring to accomodate Cython.
  * Different balance between bounds calculation and collision detection.  
    The former, which was the major load, is twice as fast while the latter is slightly slower - an overall of about X1.5.

* [draw_bounds] is now a NumericProperty (0 for False, 1 for bounds only, 2 for bounds & bounding boxes).

			
### Rotabox 0.11.0 changes

##### Added
* Method to retrieve the current coordinates of any point defined in custom_bounds:
```python
pod.center = ferris.get_point(pol index, point index)
```
* #### in Rotaboxer (bounds editor):
  * A switch to lock/unlock polygons' exporting order.  
    Before, the order changed everytime a polygon was selected.

  * An option, when dealing with an .atlas file, to export a single frame as a list or the whole as a dictionary (animated bounds).
		    
  * Ability to see and export open bounds.

##### Changed
* Some changes in documentation.

* #### in Rotaboxer (bounds editor):
  * File selectors now working properly with kivy 1.10.


### Rotabox 0.10.0 changes	

##### Added
* Two new collision detection methods, the second being similar in performance to the original one but without the need for a checking strategy.
* Open bounds support with the default collision method.

* #### in Rotaboxer (bounds editor):
  * Option to show points and polygons indices.
  * Color options for polygon drawing
	
##### Removed
* The original collision detection method is now discontinued.
* Test painting of collidable areas. No need for it anymore, since checking strategy is no longer required.

* #### in Rotaboxer (bounds editor):
  * Real-time checkpoint inspection removed. No checking strategy needed anymore.

  * Automatic checkpoint allocation removed. No checking strategy needed anymore.

  * Test painting of collidable areas removed. No checking strategy needed anymore.
  
##### Changed		
* custom_bounds' form has been simplified ([[(0, 0), (1, 0), (1, 1), (0, 1)]])

* #### in Rotaboxer (bounds editor):
  * Selecting a polygon, places it last, in the resulting list of polygons.  
    This way, when editing is done, one can determine the order of polygons in the resulting "custom_bounds" by click-selecting them, one after the other, in the required order.



### Rotabox 0.9.0 changes

##### Added	
* Touch manipulation (drag, rotation and zoom).
Since, due to the differences between the Scatter and Rotabox concepts, a way to combine the two couldn't be found, Rotabox uses the Scatter widget's code, modified to act on the actual size and position of the widget and child(essential for accurate collision detection). 
It supports single and multitouch drag, rotation and scaling (the latter two use the *origin* property in the singletouch option).

* "hidden_bounds" property.
An option for the widget to hide certain polygons from others' checks and use them as one-way detectors. A two-way check is normally required for accurate collision readings, but a second layer of bounds can have its uses where accuracy is not an issue (e.g. in longer distances, acting as the 'perception' of an enemy sprite in a game).

* "ready" BooleanProperty (False).
    Offers a Rotabox setup completion event.

* #### in Rotaboxer (bounds editor):
  * Selecting a polygon, places it first, in the resulting list of polygons.
    This way, when editing is done, one can determine the order of polygons in the resulting "custom_bounds" by click-selecting them, one after the other, in an order, opposite of the required.

  * Real-time checkpoint inspection added.

examples/_animated_bounds.py
# coin spritesheet borrowed from:
# http://www.williammalone.com/articles/html5-animation-sprite-sheet-photoshop/

from kivy.base import runTouchApp
from functools import partial
from random import random
from kivy.uix.floatlayout import FloatLayout
from kivy.uix.image import Image
from kivy.lang import Builder
from kivy.clock import Clock
import time
import sys, os
sys.path.append(os.path.abspath(".."))
from rotabox import Rotabox

Builder.load_string(str('''
<Root>:
    slotleft: slotleft
    slotright: slotright
    BoxLayout:
        id: slotleft
        size_hint: None, None
        size: root.width * .6, root.height * .01
        pos: 0, root.height * .2
        canvas:
            Color:
                rgb: 0.682, 0.251, 0.98
            Rectangle:
                pos: self.pos
                size: self.size
    BoxLayout:
        id: slotright
        size_hint: None, None
        size: root.width * .4, root.height * .01
        pos: root.width * .625, root.height * .2
        canvas:
            Color:
                rgb: 0.682, 0.251, 0.98
            Rectangle:
                pos: self.pos
                size: self.size
'''))


class Coin(Rotabox):
    def __init__(self, **kwargs):
        super(Coin, self).__init__(**kwargs)
        self.image = Image(source='atlas://coins/0')
        self.add_widget(self.image)
        self.speed = random() * .3 + 1.5
        self.custom_bounds = {'0': [[(0.201, 0.803), (0.092, 0.491),
                                     (0.219, 0.184), (0.526, 0.064),
                                     (0.83, 0.188), (0.947, 0.491),
                                     (0.83, 0.812), (0.531, 0.944)]],
                              '1': [[(0.357, 0.902), (0.17, 0.65),
                                     (0.184, 0.337), (0.343, 0.095),
                                     (0.644, 0.098), (0.813, 0.346),
                                     (0.81, 0.709), (0.651, 0.902)]],
                              '2': [[(0.402, 0.908), (0.301, 0.725),
                                     (0.298, 0.33), (0.415, 0.085),
                                     (0.627, 0.098), (0.734, 0.346),
                                     (0.73, 0.693), (0.616, 0.905)]],
                              '3': [[(0.411, 0.9), (0.364, 0.493),
                                     (0.419, 0.095), (0.595, 0.088),
                                     (0.644, 0.493), (0.588, 0.908)]],
                              '4': [[(0.422, 0.906), (0.426, 0.085),
                                     (0.572, 0.085), (0.572, 0.915)]],
                              '5': [[(0.422, 0.098), (0.592, 0.098),
                                     (0.644, 0.487), (0.568, 0.908),
                                     (0.422, 0.915), (0.364, 0.497)]],
                              '6': [[(0.395, 0.096), (0.591, 0.093),
                                     (0.701, 0.304), (0.709, 0.622),
                                     (0.595, 0.904), (0.371, 0.911),
                                     (0.262, 0.626), (0.277, 0.293)]],
                              '7': [[(0.354, 0.085), (0.612, 0.094),
                                     (0.789, 0.312), (0.811, 0.654),
                                     (0.617, 0.902), (0.341, 0.897),
                                     (0.164, 0.641), (0.187, 0.295)]],
                              '8': [[(0.305, 0.103), (0.68, 0.12),
                                     (0.87, 0.368), (0.843, 0.748),
                                     (0.582, 0.918), (0.296, 0.893),
                                     (0.092, 0.671), (0.083, 0.329)]],
                              '9': [[(0.291, 0.09), (0.68, 0.115),
                                     (0.884, 0.372), (0.893, 0.641),
                                     (0.658, 0.889), (0.305, 0.889),
                                     (0.051, 0.637), (0.06, 0.325)]]}
        # self.draw_bounds = True
        self.turn()

    def turn(self, frame=0, *args):
        if frame > 9:
            frame = 0
        try:
            self.image.source = 'atlas://coins/{}'.format(frame)
        except IndexError:
            return
        else:
            Clock.schedule_once(partial(self.turn, frame + 1), .15)


class Root(FloatLayout):
    def __init__(self, **kwargs):
        super(Root, self).__init__(**kwargs)
        self.coins = []
        self.lastTime = time.clock()
        Clock.schedule_interval(self.update, .016)

    def update(self, *args):
        if time.clock() - self.lastTime > 1:
            self.coins.append(Coin(scale=.7,
                                   pos=(self.width * .01, self.height * .2)))
            self.add_widget(self.coins[-1])
            self.lastTime = time.clock()

        for coin in self.coins:
            if coin.x > self.width * 1.1 or coin.y < self.height * -.1:
                self.remove_widget(self.coins.pop(self.coins.index(coin)))
            if (coin.collide_widget(self.slotleft)
                    or coin.collide_widget(self.slotright)):
                if self.slotleft.top > coin.center_y:
                    coin.y -= coin.speed
                    continue
                coin.x += coin.speed
                if coin.y < self.height * .19:
                    coin.y += 1
            else:
                coin.y -= coin.speed


if __name__ == '__main__':
    runTouchApp(Root())